File: cognee-frontend/src/modules/datasets/getDatasetData.ts
import { CogneeInstance } from "../instances/types";

/**
 * The server caps `/v1/datasets/{id}/data` at 100 items unless asked otherwise,
 * so a request without a limit returns a page, not the dataset. Anything that
 * needs a total must ask for one — see `getDatasetDataCount`.
 */
export const DEFAULT_DATASET_DATA_LIMIT = 100;

interface DatasetDataOptions {
  limit?: number;
  offset?: number;
  signal?: AbortSignal;
}

export default function getDatasetData(
  datasetId: string,
  instance: CogneeInstance,
  { limit = DEFAULT_DATASET_DATA_LIMIT, offset = 0, signal }: DatasetDataOptions = {},
) {
  const query = new URLSearchParams({ limit: String(limit), offset: String(offset) });

  return instance.fetch(`/v1/datasets/${datasetId}/data?${query}`, { signal })
      .then(async (response) => {
        if (!response.ok) throw new Error(`Could not load documents (${response.status})`);
        const body = await response.json();
        if (!Array.isArray(body)) throw new Error("Unexpected dataset data response");
        return body;
      });
}

/**
 * Number of documents in a dataset.
 *
 * Callers used to read `length` off the full row set to get this, which meant
 * transferring every row — tens of megabytes on a large dataset — to produce
 * one integer.
 */
export function getDatasetDataCount(
  datasetId: string,
  instance: CogneeInstance,
  signal?: AbortSignal,
): Promise<number> {
  return instance.fetch(`/v1/datasets/${datasetId}/data/count`, { signal })
      .then(async (response) => {
        if (!response.ok) throw new Error(`Could not count documents (${response.status})`);
        const body = await response.json();
        if (!Number.isSafeInteger(body?.count) || body.count < 0) {
          throw new Error("Unexpected dataset count response");
        }
        return body.count;
      });
}

/** Explicit full traversal for callers that need every document, such as schema regeneration. */
export async function getAllDatasetData(datasetId: string, instance: CogneeInstance) {
  const rows = [];
  for (;;) {
    const page = await getDatasetData(datasetId, instance, { limit: 1000, offset: rows.length });
    rows.push(...page);
    if (page.length < 1000) return rows;
  }
}

File: cognee/modules/data/methods/get_dataset_data.py
from uuid import UUID

from sqlalchemy import func, select

from cognee.infrastructure.databases.relational import get_relational_engine
from cognee.modules.data.models import Data


async def get_dataset_data(
    dataset_id: UUID,
    limit: int | None = None,
    offset: int = 0,
) -> list[Data]:
    """Return a dataset's data rows.

    `limit` defaults to None -- every row -- because most callers here are
    internal pipeline code that genuinely needs the whole set (cognify,
    sync, the estimator, permission fan-out). The HTTP route is the caller
    that must not do that, and it passes a bounded limit of its own.
    """
    db_engine = get_relational_engine()

    query = select(Data).filter(Data.dataset_id == dataset_id).order_by(Data.data_size.desc())

    if offset:
        query = query.offset(offset)
    if limit is not None:
        query = query.limit(limit)

    async with db_engine.get_async_session() as session:
        result = await session.execute(query)

        data = list(result.scalars().all())

        return data


async def count_dataset_data(dataset_id: UUID) -> int:
    """Count a dataset's data rows without fetching them.

    Callers that only need "how many documents" -- dataset list badges,
    upload/ingest polling -- used to read len() off the full row set, which
    on a large dataset meant transferring tens of megabytes to produce one
    integer. This is served by the existing dataset_id index instead.
    """
    db_engine = get_relational_engine()

    async with db_engine.get_async_session() as session:
        result = await session.execute(
            select(func.count()).select_from(Data).filter(Data.dataset_id == dataset_id)
        )

        return result.scalar_one()
